cheroot/server.py

```diff
@@ -341,16 +342,92 @@ def __next__(self):
     next = __next__
 
 
-class KnownLengthRFile:
+class RFile:
+    def __init__(self, rfile, wfile, conn):
+        self.rfile = rfile
+        self.wfile = wfile
+        self.conn = conn
+
+    def _send_100_if_needed(self):
+        if self.conn.they_are_waiting_for_100_continue:
+            mini_headers = ()
+            go_ahead = h11.InformationalResponse(status_code=100, headers=mini_headers)
+            bytes_out = self.conn.send(go_ahead)
+            self.wfile.write(bytes_out)
+
+    def _get_event(self):
+        return self.conn.next_event()
+
+    def _handle_input(self, data):
+        self.conn.receive_data(data)
+        evt = self._get_event()
+        assert isinstance(evt, h11.Data)
+        return bytes(evt.data)
+
+    def read(self, size=None):
+        raise NotImplementedError
+
+    def readline(self, size=None):
+        raise NotImplementedError
+
+    def readlines(self, sizehint=None):
+        raise NotImplementedError
+
+    def close(self):
+        self.rfile.close()
+
+
+class HyperRFile(RFile):
+    def __init__(self, rfile):
+        super(HyperRFile, self).__init__(rfile)
+
+    def read(self, size=None):
+        if self.conn.their_state != h11.SEND_BODY or size == 0:
+            return b""
+        event = self._get_event()
+        if event is h11.NEED_DATA:
+            self.conn.receive_data(self.rfile.rfile.read(size))
+            event = self._get_event()
+        if isinstance(event, h11.Data):
+            return event.data
+        else:
+            return b""
+
+    def readline(self, size=None):
+        if self.conn.their_state != h11.SEND_BODY or size == 0:
+            return b""
+        event = self._get_event()
+        if event is h11.NEED_DATA:
+            self.conn.receive_data(self.rfile.rfile.readline(size))
+            event = self._get_event()
+        assert isinstance(event, h11.Data)
+        return event.data
+
+    def readlines(self, sizehint=None):
+        if self.conn.their_state != h11.SEND_BODY:
+            return b""
+        lines = []
+        line = self.rfile.rfile.readline()
+        while line:
+            self.conn.receive_data(line)
+            event = self._get_event()
+            assert isinstance(event, h11.Data)
+            lines.append(event.data)
+            line = self.rfile.rfile.readline()
+
+        return lines
+
+
+class KnownLengthRFile(RFile):
     """Wraps a file-like object, returning an empty string when exhausted.
 
     :param rfile: ``file`` of a known size
     :param int content_length: length of the file being read
     """
 
-    def __init__(self, rfile, content_length):
+    def __init__(self, h_conn, rfile, wfile, content_length):
```

Comment:
this signature change seems to be breaking calls in other parts of the code base

Reply:
As far as I can tell, the only usages in the project besides the ones that I have modified to work with it is the ones in the old HTTPRequest implementation that I intend to remove when this PR is closer to production-ready.

Reply:
Check https://lgtm.com/projects/g/cherrypy/cheroot/rev/pr-61d0007190a3d7e76102eb80c05d5b57af1e4547. Did you take inheritance into account?

Reply:
Oops, left that in as well. The HyperRFile has been removed in favor of modifying the other existing implementations. I have just updated the repository to remove it :)